internal/controller/logger_manager.go

```diff
@@ -0,0 +1,267 @@
+		client:   client,
+		conn:     conn,
+		buffer:   make([]zapcore.Entry, 0, maxBufferSize),
+		done:     make(chan struct{}),
+		lostLogs: false,
+	}
+	go bws.run()
+	return bws
+}
+
+// Sync flushes buffered log data into grpc stream if possible.
+func (b *BufferedGrpcWriteSyncer) Sync() error {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	b.flush()
+	close(b.done)
+	return b.conn.Close()
+}
+
+// flush will attempt to dump buffer into GRPC stream if available
+func (b *BufferedGrpcWriteSyncer) flush() {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	if len(b.buffer) == 0 || b.conn == nil || b.conn.GetState() != connectivity.Ready {
+		if len(b.buffer) > maxBufferSize {
+			b.lostLogs = true
+			b.lostBytes += maxBufferSize - len(b.buffer)
+		}
+		return
+	}
+
+	for _, logEntry := range b.buffer {
+		if err := b.sendLog(logEntry); err != nil {
+			b.logger.Errorw("Failed to send log",
+				"error", err,
+				zap.Any("logBufferSize", len(b.buffer)),
+			)
+			b.lostLogs = true
+			b.lostBytes += len(logEntry.Message)
+			return
+		}
+	}
+	b.buffer = b.buffer[:0]
+
+	// If previously logs were lost, send a log message to indicate that
+	if b.lostLogs {
+		lostLogMessage := zapcore.Entry{
+			Level:   zapcore.ErrorLevel,
+			Time:    time.Now(),
+			Message: b.lostLogMessage(),
+		}
+		err := b.sendLog(lostLogMessage)
+		if err != nil {
+			b.logger.Warn("Error while sending log",
+				"error", err,
+			)
+		}
+		b.lostLogs = false
+		b.lostBytes = 0
+	}
+
+}
+
+// run flushes the buffer at the configured interval until Stop is called.
+func (b *BufferedGrpcWriteSyncer) run() {
+	ticker := time.NewTicker(flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			b.flush()
+		case <-b.done:
+			return
+		}
+	}
+}
+
+// sendLog sends the log as a string to the server.
+func (b *BufferedGrpcWriteSyncer) sendLog(logEntry zapcore.Entry) error {
+	buf, err := b.encoder.EncodeEntry(logEntry, nil)
+	if err != nil {
+		return err
+	}
+	err = b.client.Send(&pb.SendLogsRequest{
+		Request: &pb.SendLogsRequest_LogEntry{
+			LogEntry: &pb.LogEntry{
+				Level:       pb.LogLevel(logEntry.Level),
+				Time:        timestamppb.New(logEntry.Time),
+				JsonMessage: buf.String(),
+			},
+		},
+	})
+	return err
+}
+
+// UpdateClient will update BufferedGrpcWriteSyncer with new client stream and GRPC connection
+func (b *BufferedGrpcWriteSyncer) UpdateClient(client pb.KubernetesInfoService_SendLogsClient, conn *grpc.ClientConn) {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	b.client = client
+	b.conn = conn
+
+	// If logs were lost, send a log message about it
+	if b.conn.GetState() == connectivity.Ready && b.lostLogs {
+		lostLogMessage := zapcore.Entry{
+			Level:   zapcore.ErrorLevel,
+			Time:    time.Now(),
+			Message: b.lostLogMessage(),
+		}
+		err := b.sendLog(lostLogMessage)
+		if err != nil {
+			b.logger.Warn("Error while sending log",
+				"error", err,
+			)
+		}
+		b.lostLogs = false
+		b.lostBytes = 0
+	}
```

Comment:
Remove this. That should already be taken care of at the bottom of `flush`.
You should call `flush`, though, in case there are currently buffered entries.
```suggestion
	b.flush()
```